compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp

```diff
@@ -840,6 +840,78 @@ struct DotToMul : public OpRewritePattern<mhlo::DotOp> {
   }
 };
 
+// Similar to DotIsMul, this finds the case where a canonical dot general
+// can be represented using a mul operation. This includes possibly making
+// an implicit cast explicit prior the mul.
+struct DotGeneralIsMul : public OpRewritePattern<mhlo::DotGeneralOp> {
+  using OpRewritePattern<mhlo::DotGeneralOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mhlo::DotGeneralOp op,
+                                PatternRewriter &rewriter) const override {
+    auto lhs = op.lhs();
+    auto rhs = op.rhs();
+    auto lhsTy = lhs.getType().dyn_cast<RankedTensorType>();
+    auto rhsTy = rhs.getType().dyn_cast<RankedTensorType>();
+    auto resultTy = op.getType().dyn_cast<RankedTensorType>();
+
+    if (!lhsTy || !rhsTy || !resultTy) return failure();
+
+    auto dNums = op.dot_dimension_numbers();
+    auto batchDimsL = dNums.getLhsBatchingDimensions();
+    auto batchDimsR = dNums.getRhsBatchingDimensions();
+    auto contractDimsL = dNums.getLhsContractingDimensions();
+    auto contractDimsR = dNums.getRhsContractingDimensions();
+
+    // Check there are canonical number of dimensions.
+    if (batchDimsL.size() != 1 || batchDimsR.size() != 1 ||
+        contractDimsL.size() != 1 || contractDimsR.size() != 1)
+      return failure();
+
+    // Check the dimensions are the valid members.
+    if (batchDimsL.front() != 0 || batchDimsR.front() != 0 ||
+        contractDimsL.front() != 2 || contractDimsR.front() != 1)
+      return failure();
+
+    // Check that the contraction dimension is degenerate.
+    if (lhsTy.getDimSize(2) != 1) return failure();
+    if (rhsTy.getDimSize(1) != 1) return failure();
+
+    // Determine the output size of the result.
+    auto batchSize = rewriter.create<mhlo::GetDimensionSizeOp>(
```

Comment:
Can we nest the rewriter in an implicit loc op builder?

Reply:
Done.